scheduler.py

```diff
@@ -1,220 +1,368 @@
+
+                with open(sm.ens_filename, "w", encoding="utf-8") as f_out:
+                    json.dump(sm.ens, f_out, indent=4)
+
+                f_out.close()
+
+                # someRefToSleepTimer.sleep(sleep_time)
+
+            else:
+                # Python sleep with full time, sleep timer is offline
+                # we sleep here then change to ERROR to report sleep timer problem
+                logging.info(f"Calling sleep timer for {str(self.available_sleep_time)} seconds")
+                time.sleep(self.available_sleep_time) # Wait using Python sleep
+                sm.err_code = TIMER_OFFLINE
+                sm.state = ERROR()
+
         else:
-            '''
-            TODO: Wesley
+            # if not enough time, call python sleep timer then
+            # transition to CHECK_TIME state
+            logging.info(f"Calling sleep timer for {str(self.available_sleep_time)} seconds")
+            time.sleep(self.available_sleep_time) # Wait using Python sleep
+            logging.info(f"Changing state to CHECK_TIME")
+            sm.state = CHECK_TIME()
+
+
+class ERROR(State):
+    '''
+    ERROR reports errors and then sends the control back where
+    it belongs, usually to ITERATE
+    '''
+
+    def __init__(self):
+        self.err_msgs = ["No error recorded\n",                                          # 0
+                         "No active_ensembles.json file found. Unable to continue.\n",   # 1
+                         "Skipping past missed ensemble.\n",                             # 2
+                         "Hardware timer unresponsive. Defaulting to software sleep.\n"] # 3
+
+
+    def process(self, sm):
+        # Logs errors, could output to file also?
+        # can also call ERROR if we skip ensembles if we can to log them
+        logging.info(f"Running ERROR process func")
+
+        now = time.localtime()
+        curr_time_seconds = hms_to_seconds(now.tm_hour, now.tm_min, now.tm_sec)
+
+        if sm.err_code == NO_ENS_FILE:
+            logging.error(self.err_msgs[NO_ENS_FILE])
+        elif sm.err_code == MISSED_ENS:
+            logging.info(f"Missed ensemble: {sm.ens_list[sm.ens_index]['title']}")
+            logging.info(f"Current time: {str(curr_time_seconds)}")
+            logging.info(f"Ensemble target time: {str(sm.ens_list[sm.ens_index]['start_time'])}")
+            logging.error(self.err_msgs[MISSED_ENS])
+        elif sm.err_code == TIMER_OFFLINE:
+            logging.error(self.err_msgs[TIMER_OFFLINE])
+
+    def update(self, sm):
+        logging.info(f"Running ERROR update func")
+
+        if sm.err_code == NO_ENS_FILE:
+            sys.exit()
+        elif sm.err_code == MISSED_ENS:
+            sm.state = ITERATE()
+        elif sm.err_code == TIMER_OFFLINE:
+            # the SLEEP state handles the software sleep so ERROR
+            # just logs the error then passes back to ITERATE
+            sm.state = ITERATE()
```

Comment:
Discrepancy which I missed previously: If the timer is offline, after calling `time.sleep()` we proceed to the iterate state, but, if the timer is online and we just don't have enough time to shut down, we proceed to check time state after sleeping. We should be proceeding to `CHECK_TIME` for the timer offline error, since the timer being offline isn't a reason to skip an ensemble.